src/charm.py

```diff
@@ -120,11 +168,55 @@ def __init__(self, *args):
         observe(ipu_events.data_provided, self._handle_ingress_data_provided)
         observe(ipu_events.data_removed, self._handle_ingress_data_removed)
 
-        observe(self.traefik_route.on.ready, self._handle_traefik_route_ready)
+        ipr_events = self.traefik_route.on
+        observe(ipr_events.ready, self._handle_traefik_route_ready)
+        observe(ipr_events.data_removed, self._handle_ingress_data_removed)
 
         # Action handlers
         observe(self.on.show_proxied_endpoints_action, self._on_show_proxied_endpoints)
 
+    def _on_install(self, event) -> None:
+        # Generate key without a passphrase as traefik does not support it
+        # https://github.com/traefik/traefik/pull/6518
+        private_key = generate_private_key()
+        self._stored.private_key = private_key.decode()
+        # FIXME this (occasionally?!) doesn't survive a charm upgrade, and results in:
```

Comment:
Despite the docs in the library, the [Juju docs](https://juju.is/docs/sdk/install-event) are pretty clear about the caveats here. This may be much better as `on_start` -- check whether `self._stored.private_key` evaluates to true, and if it doesn't, generate a key.

Reply:
Sure, we can test that, but from the doc:

* Typically, **operations performed on install should also be considered for [upgrade-charm](https://juju.is/docs/sdk/upgrade-charm-event).**
* In some cases, [config-changed](https://juju.is/docs/sdk/config-changed-event) can be used instead of install and upgrade-charm because it is guaranteed to fire after both.

It seems to be "expected" that install event may only fire once. It's also possible that the relation join event is simply triggering first. 

* The events in `on_install` may simply be better in `on_start` (and maybe also in `on_upgrade_charm`)
* The checks in `on_certificates_relation_joined` should guard against the value being `None`
* The charm should check whether a private key already exists _before_ blindly generating one (which may change)
* And, really, we may just want to move the whole thing to the Juju Secrets API. Those patches are in OF now.

Lastly, the guess in the comments (about using Juju storage) is probably a case of trying to guess unknowable runtime state from a more or less stateless workflow (since the charm doesn't know which events have already run since startup). In the cases where you upgrade and you see this, what's the ordering of events?

`traefik-k8s` also doesn't say _not_ to use Juju storage. [If it's not set](https://github.com/canonical/operator/blob/3d0018019988a7a81470466b0bd8786835d46be3/ops/main.py#L399), OF tries to figure it out for you. If local storage has never been used (likely true), `"kubernetes"` is not in the series (true here -- no series is set), and the controller version is new enough (also true), `use_juju_for_storage` will default to `True`. We probably _are_ already using it, and that can be verified by looking for [this log message](https://github.com/canonical/operator/blob/3d0018019988a7a81470466b0bd8786835d46be3/ops/main.py#L352).

What events are being fired?

Reply:
> `use_juju_for_storage` will default to `True`

Good point.

I also haven't encountered this error since I put in the comment.
I'll try to write a few utests around this and we'll take it from there.

Reply:
Confirmed that we need to manually set `use_juju_for_storage=True` (see https://github.com/canonical/operator/issues/880).